Modernize circle, include github publishing

dsoo opened the pull request:
Similar to what was just done for libhoney-js:

* Updates to the CircleCI build process so that it more closely matches our "state of the art" process, including:
    * Matrix for Ruby versions
    * Publish to github releases in addition to Rubygems.
* Added CHANGELOG.md

martin308 commented on .circleci/config.yml:
```diff
@@ -1,66 +1,116 @@
+matrix_rubyversions: &matrix_rubyversions
+  matrix:
+    parameters:
+      rubyversion: ["2.3", "2.4", "2.5", "2.6", "2.7"]
```
ooh could you please add 2.2? We say we support it in the gemspec 🤫

dsoo replied:
Done!

dsoo commented:
@martin308 - Okay, this is ready to go now.

martin308 commented on .circleci/config.yml:
```diff
@@ -83,27 +125,24 @@ workflows:
+          <<: *filters_always
+          <<: *matrix_rubyversions
+      - build_artifacts:
+          <<: *filters_always
```
should this be using the publish filters? It doesn't seem like we need the artifacts unless we are publishing to GitHub?

```suggestion
          <<: * filters_publish
```

dsoo replied:
I think it's handy to have the built artifact, it allows us to build and provide gems for manual installations for "unofficial" releases, e.g. to test/cut a bug fix patch for a customer without having to create an official release.

martin308 replied:
hmm ok, perhaps we should build it with a different version then? It might be a bit weird to have version `x.x.x` published on ruby gems being different to the same version that we give a customer

dsoo replied:
Oh - we are building it, but not publishing it in the current implementation. You will see the artifact in the CI artifacts, but not on github or rubygems.

martin308 commented on .circleci/config.yml:
```diff
@@ -83,27 +125,24 @@ workflows:
+          <<: *filters_publish
+          context: Honeycomb Secrets
+          requires:
+            - test
```
should this step be changed to require `build_artifacts` and then push the built gem instead?

dsoo replied:
Yeah, I will change this.

martin308 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,9 @@
+# libhoney-rb changelog
+
+## 1.14.6
```
is the current version `1.14.7`?

dsoo replied:
Ah, we cut a new version while this branch was sitting while I was on vacation. I can update.

martin308 reviewed (commented):
looks like some good improvements, a couple of thoughts inline

dsoo commented:
@martin308 Okay, take N. Updated github repo config to include our CI user, and tested the full build including tagging and publishing. I think I've addressed all of the outstanding issues.

martin308 commented on lib/libhoney/version.rb:
```diff
@@ -1,3 +1,3 @@
 module Libhoney
-  VERSION = '1.14.7'.freeze
+  VERSION = '1.14.7-beta'.freeze
```
what's the story here?

dsoo replied:
This was due to testing the version bumping tooling and testing package publishing - I wish we had a better way of testing that. I'll roll this back before I merge.

martin308 reviewed (approved):
Just one question on the beta version being checked in
